sw/device/tests/flash_ctrl_ops_test.c

```diff
@@ -0,0 +1,393 @@
+      plic, kTopEarlgreyPlicIrqIdFlashCtrlOpDone, kTopEarlgreyPlicTargetIbex0,
+      kDifToggleEnabled));
+}
+
+/**
+ * Compares the expected and fired IRQs and clears both.
+ */
+static void compare_and_clear_irq_variables(void) {
+  for (int i = 0; i < FLASH_CTRL_NUM_IRQS; ++i) {
+    CHECK(expected_irqs[i] == fired_irqs[i], "expected IRQ mismatch = %0d", i);
+  }
+  clear_irq_variables();
+}
+
+/**
+ * Tests the interrupts for erase, write and
+ * read of the specified information partition.
+ * Confirms that the written data is read back correctly.
+ */
+static void do_info_partition_test(uint32_t partition_number,
+                                   const uint32_t *test_data) {
+  uint32_t address = flash_ctrl_testutils_info_region_setup(
+      &flash_state, partition_number, kFlashInfoBank, kPartitionId);
+
+  CHECK_DIF_OK(dif_flash_ctrl_set_prog_fifo_watermark(&flash_state, 1));
+  CHECK_DIF_OK(dif_flash_ctrl_set_read_fifo_watermark(&flash_state, 8));
+
+  clear_irq_variables();
+
+  expected_irqs[kDifFlashCtrlIrqOpDone] = true;
+  CHECK(flash_ctrl_testutils_erase_page(&flash_state, address, kPartitionId,
+                                        kDifFlashCtrlPartitionTypeInfo) == 0);
+  compare_and_clear_irq_variables();
+
+  expected_irqs[kDifFlashCtrlIrqOpDone] = true;
+  expected_irqs[kDifFlashCtrlIrqProgEmpty] = true;
+  expected_irqs[kDifFlashCtrlIrqProgLvl] = true;
+  CHECK(flash_ctrl_testutils_write_page(
+            &flash_state, address, kPartitionId, test_data,
+            kDifFlashCtrlPartitionTypeInfo, kInfoSize) == 0);
+
+  compare_and_clear_irq_variables();
+
+  uint32_t readback_data[kInfoSize];
+  expected_irqs[kDifFlashCtrlIrqOpDone] = true;
+  expected_irqs[kDifFlashCtrlIrqRdLvl] = true;
+  expected_irqs[kDifFlashCtrlIrqRdFull] = true;
+  CHECK(flash_ctrl_testutils_read_page(
+            &flash_state, address, kPartitionId, readback_data,
+            kDifFlashCtrlPartitionTypeInfo, kInfoSize, 1) == 0);
+
+  compare_and_clear_irq_variables();
+
+  CHECK_BUFFER(readback_data, test_data, kInfoSize);
+}
+
+/**
+ * Tests the interrupts for read of bank0 data partition.
+ * Only read is tested as this partition contains the program
+ * code so should not be erased or written.
+ * No confirmation of the read back data validity is performed.
+ */
+static void do_bank0_data_partition_test(void) {
+  uint32_t address = flash_ctrl_testutils_data_region_setup(
+      &flash_state, kRegionBaseBank0Page0Index, kFlashBank0DataRegion,
+      kRegionSize);
+
+  CHECK_DIF_OK(dif_flash_ctrl_set_read_fifo_watermark(&flash_state, 8));
+
+  clear_irq_variables();
+  expected_irqs[kDifFlashCtrlIrqOpDone] = true;
+  expected_irqs[kDifFlashCtrlIrqRdLvl] = true;
+  expected_irqs[kDifFlashCtrlIrqRdFull] = true;
+  uint32_t readback_data[kDataSize];
+  CHECK(flash_ctrl_testutils_read_page(
+            &flash_state, address, kPartitionId, readback_data,
+            kDifFlashCtrlPartitionTypeData, kDataSize, 1) == 0);
```

Comment:
Is this via controller or via the host path? Recommend checking the read data via host path.

Reply:
it would be good to check both honestly, to make sure there are no discrepancies on either.

Reply:
The program / read-checks via the controller go over the same TL interface, which are well exercised at the block level. At chip level, the goal is to maximize the usage of the block's IOs, so I suggested read-checking via the host path instead - that way we exercise 2 separate flash ctrl TL interfaces in the same test.